ModulemdTranslationHelpers/Utils.py

```diff
@@ -11,6 +11,100 @@
+    Get the most-recently built versions of each (module,stream) pair from
+    a Koji tag
+    :param session: A Koji session
+    :param tag: A koji tag
+    :return: A list of the most recent build of all modules in the tag.
+    """
+
+    # Koji sometimes disconnects for no apparent reason. Retry up to 5
+    # times before failing.
+    for attempt in range(5):
+        try:
+            tagged = session.listTagged(tag)
+        except requests.exceptions.ConnectionError:
+            if debug:
+                print("Connection lost while retrieving builds for tag %s, "
+                      "retrying..." % tag,
+                      file=sys.stderr)
+        else:
+            # Succeeded this time, so break out of the loop
+            break
+
+    # Find the latest, in module terms.  Pungi does this.
+    # Collect all contexts that share the same NSV.
+    NSVs = {}
+    for entry in tagged:
+        name, stream = entry['name'], entry['version']
+        version = entry['release'].rsplit('.', 1)[0]
+
+        NSVs[name] = NSVs.get(name, {})
+        NSVs[name][stream] = NSVs[name].get(stream, {})
+        NSVs[name][stream][version] = NSVs[name][stream].get(version, [])
+        NSVs[name][stream][version].append(entry)
+
+    latest = []
+    for name in NSVs:
+        for stream in NSVs[name]:
+            version = sorted(list(NSVs[name][stream].keys()))[-1]
+            latest.extend(NSVs[name][stream][version])
+
+    return latest
+
+
+def get_index_from_tags(session, tags, debug=False):
+    """
+    Construct a ModuleIndex object from the contents of the provided tags.
+    :param session: A Koji session
+    :param tags: A set of Koji tags from which module metadata should be pulled
+    :param debug: Whether to print debugging information to the console
+    :return: A ModuleIndex object. Raises an exception if any of the
+    retrieved modulemd is invalid.
+    """
+
+    tagged_builds = []
+    for tag in tags:
+        tagged_builds.extend(get_latest_modules_in_tag(session, tag, debug))
+
+    # Make the list unique since some modules may have multiple tags
+    unique_builds = {}
+    for build in tagged_builds:
+        unique_builds[build['id']] = build
+
+    for build_id in unique_builds.keys():
+        # Koji sometimes disconnects for no apparent reason. Retry up to 5
+        # times before failing.
+        for attempt in range(5):
+            try:
+                build = session.getBuild(build_id)
+            except requests.exceptions.ConnectionError:
+                if debug:
+                    print("Connection lost while processing buildId %s, "
+                          "retrying..." % build_id,
+                          file=sys.stderr)
+            else:
+                # Succeeded this time, so break out of the loop
+                break
+        if debug:
+            print("Processing %s:%s" % (build['package_name'], build['nvr']))
+
+        modulemds = Modulemd.objects_from_string(
```

Comment:
I'm assuming this function is to retrieve a YAML document. But I don't know if this will work. Also, the `getBuild()` method returns a slightly different set of values. The available values as per the documentation is: 
```
A map will be returned containing the following keys*:
      id: build ID
      package_id: ID of the package built
      package_name: name of the package built
      name: same as package_name
      version
      release
      epoch
      nvr
      state
      task_id: ID of the task that kicked off the build
      owner_id: ID of the user who kicked off the build
      owner_name: name of the user who kicked off the build
      volume_id: ID of the storage volume
      volume_name: name of the storage volume
      creation_event_id: id of the create_event
      creation_time: time the build was created (text)
      creation_ts: time the build was created (epoch)
      start_time: time the build was started (may be null)
      start_ts: time the build was started (epoch, may be null)
      completion_time: time the build was completed (may be null)
      completion_ts: time the build was completed (epoch, may be null)
      source: the SCM URL of the sources used in the build
      extra: dictionary with extra data about the build
```

What values should I be using now?

Reply:
The modulemd YAML is all under the `extra` key and the existing way of accessing it (`build['extra']['typeinfo']['module']['modulemd_str']`) should be correct.

This used the old libmodulemd1 approach of getting the YAML, but it's completely out of date. You can actually simplify this immensely by just creating the index before this loop and then calling `index.update_from_string(build['extra']['typeinfo']['module']['modulemd_str'])` multiple times through the loop. Each `update_from_string()` will add the found content to the `ModulemdIndex`.

I think it's probably safe to just drop the `len(modulemds)` check, since it should be impossible and it will simplify things a lot here.

Reply:
Ohhh I got it now. 
`build['extra']['typeinfo']['module']['modulemd_str']` was accessing nested components of `extra`. I misunderstood them all as `keys` of the build.